pkgs/applications/version-management/sapling/default.nix

```
@@ -0,0 +1,166 @@
+  # to src.
+  cargoDeps = rustPlatform.importCargoLock {
+    lockFile = ./Cargo.lock;
+    outputHashes = {
+      "cloned-0.1.0" = "sha256-c3CPWVjOk+VKBLD6WuaYZvBoKi5PwgXmiwxKoCk0bsI=";
+      "deltae-0.3.0" = "sha256-a9Skaqs+tVTw8x83jga+INBr+TdaMmo35Bf2wbfR6zs=";
+      "fb303_core-0.0.0" = "sha256-yoKKSBwqufFayLef2rRpX5oV1j8fL/kRkXBXIC++d7Q=";
+      "fbthrift-0.0.1+unstable" = "sha256-jtsDE5U/OavDUXRAE1N8/nujSPrWltImsFLzHaxfeM0=";
+      "reqwest-0.11.11" = "sha256-uhc8XhkGW22XDNo0qreWdXeFF2cslOOZHfTRQ30IBcE=";
+      "serde_bser-0.3.1" = "sha256-KCAC+rbczroZn/oKYTVpAPJl40yMrszt/PGol+JStDU=";
+    };
+  };
+  postPatch = ''
+    cp ${./Cargo.lock} Cargo.lock
+  '';
+
+  # TODO: Automate Sapling build dep fetching
+  # Since the derivation builder doesn't have network access to remain pure,
+  # fetch the artifacts manually and link them. Then replace the hardcoded URLs
+  # with filesystem paths for the curl calls
+  postUnpack = ''
+    mkdir $sourceRoot/fetched_build_deps
+    ( cd $sourceRoot/fetched_build_deps
+      ln -s ${ptyprocess} ptyprocess-0.7.0-py2.py3-none-any.whl
+      ln -s ${pexpect} pexpect-4.8.0-py2.py3-none-any.whl
+      ln -s ${ipython} ipython-7.16.1-py3-none-any.whl
+      ln -s ${simplegeneric} simplegeneric-0.8.1.zip
+      ln -s ${toml} toml-0.10.2-py2.py3-none-any.whl
+      ln -s ${colorama} colorama-0.4.4-py2.py3-none-any.whl
+      ln -s ${backcall} backcall-0.2.0-py2.py3-none-any.whl
+      ln -s ${setuptools} setuptools-57.0.0-py3-none-any.whl
+      ln -s ${wcwidth} wcwidth-0.2.5-py2.py3-none-any.whl
+      ln -s ${prompt_toolkit} prompt_toolkit-2.0.10-py3-none-any.whl
+      ln -s ${decorator} decorator-5.0.9-py3-none-any.whl
+      ln -s ${pickleshare} pickleshare-0.7.5-py2.py3-none-any.whl
+      ln -s ${Pygments} Pygments-2.9.0-py3-none-any.whl
+      ln -s ${traitlets} traitlets-4.3.3-py2.py3-none-any.whl
+      ln -s ${six} six-1.16.0-py2.py3-none-any.whl
+      ln -s ${ipython_genutils} ipython_genutils-0.2.0-py2.py3-none-any.whl
+      ln -s ${ipdb} ipdb-0.13.9.tar.gz
+      ln -s ${Cython} Cython-0.29.26.tar.gz )
+
+    sd 'https://files.pythonhosted.org/packages/.*?/.*?/.*?/' "file://$NIX_BUILD_TOP/$sourceRoot/fetched_build_deps/" $sourceRoot/setup.py
+  '';
+
+  SAPLING_OSS_BUILD = "true";
+  HGNAME = "sl";
+  SAPLING_VERSION = version;
+  # FIXME: python -c 'import struct, hashlib; str(struct.unpack(">Q", hashlib.sha1("< SAPLING_VERSION >".encode("ascii")).digest()[:8])[0])'
+  SAPLING_VERSION_HASH = "5535144625961033752";
+
+  doCheck = false;
+
+  nativeBuildInputs = [
+    sd
+    curl
+    pkg-config
+  ] ++ (with rustPlatform; [
+    cargoSetupHook
+    rust.cargo
+    rust.rustc
+  ]);
+
+  buildInputs = [
+    openssl
+  ];
+
+  meta = with lib; {
+    description = "A Scalable, User-Friendly Source Control System";
+    homepage = "https://sapling-scm.com";
+    license = licenses.gpl2Only;
+    maintainers = with maintainers; [ pbar ];
+    platforms = platforms.unix;
```

```suggestion
    platforms = platforms.linux;
```
fails with cryptic linking error on darwin

---

It is cryptic, I don't have a darwin device to test on currently so will constrain to linux.